mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt

```diff
@@ -131,4 +131,44 @@ fun BuildSteps.runAcceptanceTests() {
             """.trimIndent()
         })
     }
+}
+
+fun BuildSteps.saveArtifactsToGCS() {
+    step(ScriptBuildStep {
+        name = "Tasks after running nightly tests: push artifacts(debug logs) to GCS"
+        scriptContent = """
+            #!/bin/bash
+            echo "Post-test step - storge artifacts(debug logs) to GCS"
+
+            # Authenticate gcloud CLI
+            echo "${'$'}{GOOGLE_CREDENTIALS}" > google-account.json
+            chmod 600 google-account.json
+            gcloud auth activate-service-account --key-file=google-account.json
+
+            # Detect Trigger Method 
+            TRIGGERED_BY_USERNAME=%teamcity.build.triggeredBy.username%
+            BRANCH_NAME=%teamcity.build.branch%
+            if [[ "${'$'}TRIGGERED_BY_USERNAME" = "n/a" ]] ; then
+                echo "Build was triggered as part of automated testing. We know this because the `triggeredBy.username` value was `n/a`, value: ${'$'}{TRIGGERED_BY_USERNAME}"
+                FOLDER="nightly"
+            else
+                echo "Build was triggered manually. We know this because `triggeredBy.username` has a non- `n/a` value: ${'$'}{TRIGGERED_BY_USERNAME}"
+                FOLDER="manual/${'$'}BRANCH_NAME"
+            fi
+
+
+            # Copy logs to GCS
+            gsutil -m cp %teamcity.build.checkoutDir%/debug* gs://teamcity-nightly-logs/%PROVIDER_NAME%/${'$'}{FOLDER}/%env.BUILD_NUMBER%/
```

Comment:
I wanted to check if files would have unique paths if someone ran tests manually on the same commit, and it looks like they wouldn't be unique.

[We set env.BUILD_NUMBER to the short commit hash](https://github.com/hashicorp/terraform-provider-google/blob/82eda93fed4a662be30ee6dc56e507322539e8a7/.teamcity/components/builds/build_steps.kt#L25-L38) (this is copied from how the old TeamCity projects worked) so this value isn't unique per build. For example, [the first artifact listed here](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ACCESSAPPROVAL/138150?buildTab=artifacts) would have the URL `gs://teamcity-nightly-logs/google/nightly/2061d55/debug-google-2061d55-TestAccAccessApprovalSettings__folder.txt`. If that commit was the tip of main branch after the last PR is merged on a Friday evening it'll be the same commit used for Saturday, Sunday and Monday's early-morning tests and the bucket would only contain the Monday's artifact.

I took a look at the `build.counter` configuration parameter as an alternative. It looks like this value increments for a given build configuration, i.e. there is a past build of bigquery tests that has counter=100, and there'll be a past build of the pubsub tests that has counter=100. This could be used to make the GCS objects unique, either by extending the prefix or by changing [the artifact's filename](https://github.com/hashicorp/terraform-provider-google/blob/82eda93fed4a662be30ee6dc56e507322539e8a7/.teamcity/components/builds/build_parameters.kt#L247) to use `%teamcity.build.counter%` instead of `%env.BUILD_NUMBER%`

As a real world example, here are some past nightly tests for AlloyDB:
- [Saturday April 13th](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ALLOYDB/132693?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildChangesSection=true), build number = 13a7e31, build count = 199
- [Sunday April 14th](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ALLOYDB/133249?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildTestsSection=true), build number = 13a7e31, build count = 200
- [Monday April 15th](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ALLOYDB/133922?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false), build number = 13a7e31, build count = 201

Reply:
That comment was a bit of a ramble, but to distill it:

- This problem would be encountered by every test every weekend, so needs addressing
- Solutions are to either change the prefix of the GCS object or change the original artifact's filename
    - the filenames of the artifacts can be changed without breaking anything, so I'd recommend that option

Reply:
Good catch! I think we could add `counter` to the artifact's filename. I also want to keep the build number in the filename as it'd clear to tell which commit the test is at.

I wonder if we can get the time/date when the build is triggered by %teamcity.build.startDate%, as I was thinking this might be useful to group logs in (cron-triggered) nightly test.

Reply:
Re: `%teamcity.build.startDate%` , I can't see that value available when I look at [the parameters etc for a random build](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ACCESSAPPROVAL/139161?buildTab=buildParameters&buildParametersTab=parameters&buildParametersShowDependencies=false). When I google about it I see [this JetBrains issue thats >10 years old](https://youtrack.jetbrains.com/issue/TW-4080/Predefined-properties-to-denote-build-start-date-and-time) 🤯 requesting the ability to access the start times for builds. The existing workaround is to install a plugin which I don't think is an option for us.


Aside: We can set values on builds ourselves using [service messages](https://www.jetbrains.com/help/teamcity/service-messages.html), and this is how the [build number is changed to the git commit](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L32) and how we [tag builds](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L50). You could make a build's first action be to tag itself with the current time, but that would only reflect when the build's user-defined steps started running and not when the build overall started running. Thought I'd share that option if it's still useful that way!

Reply:
>Good catch! I think we could add counter to the artifact's filename. I also want to keep the build number in the filename as it'd clear to tell which commit the test is at.

This sounds like a good idea!

Reply:
> Re: `%teamcity.build.startDate%` , I can't see that value available when I look at [the parameters etc for a random build](https://hashicorp.teamcity.com/buildConfiguration/TerraformProviders_GoogleCloud_GOOGLE_NIGHTLYTESTS_GOOGLE_PACKAGE_ACCESSAPPROVAL/139161?buildTab=buildParameters&buildParametersTab=parameters&buildParametersShowDependencies=false). When I google about it I see [this JetBrains issue thats >10 years old](https://youtrack.jetbrains.com/issue/TW-4080/Predefined-properties-to-denote-build-start-date-and-time) 🤯 requesting the ability to access the start times for builds. The existing workaround is to install a plugin which I don't think is an option for us.
> 
> Aside: We can set values on builds ourselves using [service messages](https://www.jetbrains.com/help/teamcity/service-messages.html), and this is how the [build number is changed to the git commit](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L32) and how we [tag builds](https://github.com/hashicorp/magic-modules/blob/6681336af2f445057526fd050033546ee6b87502/mmv1/third_party/terraform/.teamcity/components/builds/build_steps.kt#L50). You could make a build's first action be to tag itself with the current time, but that would only reflect when the build's user-defined steps started running and not when the build overall started running. Thought I'd share that option if it's still useful that way!

yeah, I also realized that the build parameters page does not contain any date related info unfortunately 😢 . I wonder if we can just simply get the run date by `current_date=$(date+"%Y-%m-%d")` in the bash script and use that value. It'd be nice if we can use date to group nightly test logs and show that in the storage path, as it'd be clear to viewers, I think.

Reply:
Ah if you're just wanting the date then that's super easy - your suggestion will work. Just bear in mind that all agents etc use UTC!